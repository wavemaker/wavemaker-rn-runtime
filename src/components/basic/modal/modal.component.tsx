import React from 'react';
import { View, Modal as ReactModal } from 'react-native';
import { BaseComponent, BaseComponentState } from '@wavemaker/app-rn-runtime/core/base.component';

import WmModalProps from './modal.props';
import { DEFAULT_CLASS, WmModalStyles } from './modal.styles';

export class WmModalState extends BaseComponentState<WmModalProps> {

}

export default class WmModal extends BaseComponent<WmModalProps, WmModalState, WmModalStyles> {

  constructor(props: WmModalProps) {
    super(props, DEFAULT_CLASS, new WmModalProps());
  }

  renderWidget(props: WmModalProps) {
    return React.createElement(View, {
      style: this.styles.root
    }, React.createElement(ReactModal, {
      animationType: props.animationType as any,
      transparent: true,
      //@ts-ignore
      style: this.styles.content
    }, props.children));
  }
}
