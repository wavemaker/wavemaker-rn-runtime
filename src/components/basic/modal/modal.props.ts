import { BaseProps } from '@wavemaker/app-rn-runtime/core/base.component';

export default class WmModalProps extends BaseProps {
    children = null as any;
    animationType? = 'slide';
}