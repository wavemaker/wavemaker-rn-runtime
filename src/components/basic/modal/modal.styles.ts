import BASE_THEME, { AllStyle }  from '@wavemaker/app-rn-runtime/styles/theme';
import { BaseStyles } from '@wavemaker/app-rn-runtime/core/base.component';


export type WmModalStyles = BaseStyles & {
    content: AllStyle,
};

export const DEFAULT_CLASS = 'app-modal';
export const DEFAULT_STYLES: WmModalStyles = {
    root: {
        backgroundColor: 'rgba(0, 0, 0, 0)'
    },
    text: {},
    content: {
        borderColor: 'rgba(0, 0, 0, 0)',
        borderWidth: 0
    }
};

BASE_THEME.addStyle('centered-modal', '', {
    root: {
        flex: 1,
        justifyContent: "center",
        alignItems: "center",
        alignSelf: 'center',
    },
    text: {},
    content: {
        alignSelf: 'center'
    }
} as WmModalStyles);


BASE_THEME.addStyle(DEFAULT_CLASS, '', DEFAULT_STYLES);