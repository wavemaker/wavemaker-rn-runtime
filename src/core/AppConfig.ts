export interface Drawer {
    setContent: (c: React.ReactNode) => void;
    getContent: () => React.ReactNode;
    setAnimation: (animation: string) => void;
    getAnimation: () => string;
}

export default interface AppConfig {
    assets: any;
    appProperties: any;
    appLocale: any;
    url: string;
    loadApp: boolean;
    refresh: (complete?: boolean) => void,
    currentPage?: any;
    pages?: any[];
    landingPage: string;
    partials?: any[];
    drawer: Drawer;
    app: any;
    setDrawerContent: any;
    theme: any;
    drawerType: any;
    getServiceDefinitions: any;
    loggedInUser: any;
    selectedLocale: string;
}