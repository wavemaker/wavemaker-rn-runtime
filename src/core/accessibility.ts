import { AccessibilityInfo } from 'react-native';
import { isAndroid, removeUndefinedKeys } from './utils';

let _isScreenReaderEnabled = false;

AccessibilityInfo.addEventListener(
  'screenReaderChanged',
  flag => {
    _isScreenReaderEnabled = flag;
  },
);

export const isScreenReaderEnabled = () => _isScreenReaderEnabled;

async function getScreenReaderStatus() {
  _isScreenReaderEnabled = await AccessibilityInfo.isScreenReaderEnabled();
}

getScreenReaderStatus();

export enum AccessibilityWidgetType {
  BUTTON = 'button',
  PICTURE = 'picture',
  TEXT = 'text',
  NUMBER = 'number',
  TEXTAREA = 'textarea',
  SELECT = 'select',
  CHIPS = 'chips',
  CURRENCY = 'currency',
  RADIOSET = 'radioset',
  CHECKBOX = 'checkbox',
  TOGGLE = 'toggle',
  SWITCH = 'switch',
  DATE = 'date',
  VIDEO = 'video',
  PROGRESSBAR = 'progressbar',
  PROGRESSCIRCLE = 'progresscircle',
  LABEL = 'label',
  ANCHOR = 'anchor',
  MESSAGE = 'message',
  SEARCH = 'search',
  ICON = 'icon',
  NAV = 'nav',
  POVOVER = 'popover',
  WEBVIEW = 'webview',
  LINECHART = 'linechart',
};

  
export type AccessibilityPropsType = {
  accessible?: boolean;
  accessibilityLabel?: string;
  accessibilityLabelledBy?: string;
  accessibilityHint?: string;
  accessibilityRole?: 'button' | 'link' | 'header' | 'search' | 'image' | 'imagebutton' | 'none' | 'summary' | 'text' | 'progressbar' | 'grid' | 'alert';
  accessibilityState?: {
    disabled?: boolean;
    selected?: boolean;
    checked?: boolean;
    expanded?: boolean;
  };
  accessibilityValue?: {
    min?: number;
    max?: number;
    now?: number;
    text?: string;
  };
  accessibilityActions?: Array<{
    name: string;
    label?: string;
  }>;
  accessibilityLiveRegion?: 'none' | 'polite' | 'assertive';
  accessibilityLanguage?: any;
  accessibilityElementsHidden?: boolean;
  accessibilityViewIsModal?: boolean;
};
  

export const getAccessibilityProps = (widgetType: AccessibilityWidgetType, accessibilityProps: AccessibilityPropsType | any) => {
  let props: AccessibilityPropsType = {accessible: true};
  if (!_isScreenReaderEnabled) {
    return {};
  }
  switch (widgetType) {
    case AccessibilityWidgetType.BUTTON:
    case AccessibilityWidgetType.TEXT:
    case AccessibilityWidgetType.NUMBER:
    case AccessibilityWidgetType.TEXTAREA:
    case AccessibilityWidgetType.SELECT:
    case AccessibilityWidgetType.CURRENCY:
    case AccessibilityWidgetType.TOGGLE:
    case AccessibilityWidgetType.DATE:
    case AccessibilityWidgetType.LABEL:
    case AccessibilityWidgetType.ANCHOR:
    case AccessibilityWidgetType.MESSAGE:    
    case AccessibilityWidgetType.SEARCH: 
    case AccessibilityWidgetType.PICTURE: 
    case AccessibilityWidgetType.ICON:
    case AccessibilityWidgetType.NAV:
    case AccessibilityWidgetType.POVOVER:
    case AccessibilityWidgetType.WEBVIEW:
    case AccessibilityWidgetType.LINECHART:  
    case AccessibilityWidgetType.VIDEO: {
      props.accessibilityLabel = accessibilityProps.accessibilitylabel || accessibilityProps.caption;
      props.accessibilityHint = accessibilityProps.hint;
      props.accessibilityRole = accessibilityProps.accessibilityrole;

      if (
        widgetType === AccessibilityWidgetType.BUTTON ||
        widgetType === AccessibilityWidgetType.TEXT ||
        widgetType === AccessibilityWidgetType.NUMBER ||
        widgetType === AccessibilityWidgetType.TEXTAREA ||
        widgetType === AccessibilityWidgetType.SELECT ||
        widgetType === AccessibilityWidgetType.TOGGLE ||
        widgetType === AccessibilityWidgetType.DATE
      ) {
        props.accessibilityState = { disabled: accessibilityProps.disabled };
      }
      if (
        (widgetType === AccessibilityWidgetType.TEXT ||
          widgetType === AccessibilityWidgetType.NUMBER ||
          widgetType === AccessibilityWidgetType.TEXTAREA ||
          widgetType === AccessibilityWidgetType.SELECT ||
          widgetType === AccessibilityWidgetType.CURRENCY ||
          widgetType === AccessibilityWidgetType.TOGGLE) &&
        isAndroid()
      ) {
        props.accessibilityLabelledBy =
          accessibilityProps.accessibilitylabelledby;
      }
      if (
        widgetType === AccessibilityWidgetType.NUMBER ||
        widgetType === AccessibilityWidgetType.CURRENCY
      ) {
        props.accessibilityValue = {
          min: accessibilityProps.minvalue,
          max: accessibilityProps.maxvalue,
        };
      }
      if (widgetType === AccessibilityWidgetType.SELECT) {
        props.accessibilityState = {
          ...props.accessibilityState,
          expanded: accessibilityProps.expanded,
        };
      }
      if (widgetType === AccessibilityWidgetType.TOGGLE) {
        props.accessibilityState = {
          ...props.accessibilityState,
          selected: accessibilityProps.selected,
        };
      }
      break;
    }

    case AccessibilityWidgetType.CHIPS: {
      props.accessibilityLabel = accessibilityProps.accessibilitylabel || accessibilityProps.caption;
      props.accessibilityHint = accessibilityProps.hint;
      props.accessibilityState = {
        disabled: accessibilityProps.disabled,
        selected: accessibilityProps.selected,
      };
      break;
    }

    case AccessibilityWidgetType.RADIOSET: {
      props.accessibilityState = {
        disabled: accessibilityProps.readonly || accessibilityProps.disabled,
        selected: accessibilityProps.selected,
      };
      break;
    }

    case AccessibilityWidgetType.CHECKBOX: {
      props.accessibilityState = {
        disabled: accessibilityProps.readonly ||  accessibilityProps.disabled,
        checked: accessibilityProps.checked,
      };
      break;
    }

    case AccessibilityWidgetType.SWITCH: {
      props.accessibilityState = {
        disabled: accessibilityProps.disabled,
        selected: accessibilityProps.selected,
      };
      break;
    }
    case AccessibilityWidgetType.PROGRESSBAR:
    case AccessibilityWidgetType.PROGRESSCIRCLE: {
      props.accessibilityLabel = accessibilityProps.accessibilitylabel || accessibilityProps.caption;
      props.accessibilityRole = accessibilityProps.accessibilityrole;
      break;
    }
    default:
      break;
  }

  const finalProps = removeUndefinedKeys(props);
  // console.log('finalProps', finalProps)

  return finalProps;
}
  