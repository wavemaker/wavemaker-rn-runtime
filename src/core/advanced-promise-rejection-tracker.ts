/**
 * Advanced Promise Rejection Tracker
 *
 * Provides comprehensive promise rejection tracking across different JavaScript engines.
 * This module addresses the critical gap where API call failures and async function errors
 * are not caught by React Error Boundaries, ensuring all promise rejections are handled.
 *
 * Engine Support:
 * - Web: Uses browser's 'unhandledrejection' event
 * - Hermes: Uses HermesInternal.enablePromiseRejectionTracker
 * - JSC: Uses React Native's 'promise/setimmediate/rejection-tracking' module
 *
 * Coverage Includes:
 * - API call failures (fetch, axios, any HTTP library)
 * - Async function errors (unhandled throws in async/await)
 * - Manual Promise.reject() calls
 * - Third-party library promise rejections
 *
 * Research Notes:
 * This implementation was developed through extensive research into React Native's
 * promise handling across different engines. Each engine has its own optimal approach:
 * - Hermes provides a native tracking API that hooks into the engine's internals
 * - JSC can use React Native's internal promise tracking module
 * - Web browsers have standardized unhandledrejection events
 *
 * The approach automatically detects the best available method and falls back
 * gracefully when native tracking is unavailable.
 */

import { Platform } from 'react-native';

// TypeScript declarations for global objects
declare global {
  interface Global {
    HermesInternal?: {
      enablePromiseRejectionTracker?: (options: {
        allRejections: boolean;
        onUnhandled: (id: number, error: any) => void;
        onHandled: (id: number) => void;
      }) => void;
    };
  }
}

export interface PromiseRejectionHandler {
  (error: any): void;
}

/**
 * Multi-engine promise rejection tracking implementation
 */
export class AdvancedPromiseRejectionTracker {
  private onRejectionHandler: PromiseRejectionHandler;
  private cleanupFunctions: (() => void)[] = [];

  constructor(onRejection: PromiseRejectionHandler) {
    this.onRejectionHandler = onRejection;
  }

  /**
   * Setup promise rejection tracking based on available APIs
   * Returns true if native tracking was enabled, false if fallback used
   */
  public setup(): boolean {
    // Automatically detect and setup the best available promise rejection tracking method

    if (Platform.OS === 'web') {
      return this.setupWebTracking();
    }

    // For mobile platforms, try native tracking methods in order of preference
    if (this.setupHermesTracking()) {
      return true;
    }

    // Fallback to React Native internal tracking for JSC
    if (this.setupReactNativeTracking()) {
      return true;
    }

    // No native tracking available - this should rarely happen in modern RN versions
    return false;
  }

  /**
   * Method 1: Web Environment - Use unhandledrejection event
   */
  private setupWebTracking(): boolean {
    try {
      if (typeof window === 'undefined' || !window.addEventListener) {
        return false;
      }

      const handler = (event: PromiseRejectionEvent) => {
        event.preventDefault(); // Prevent browser's default handling
        const error =
          event.reason instanceof Error
            ? event.reason
            : new Error(String(event.reason));
        this.onRejectionHandler(error);
      };

      window.addEventListener('unhandledrejection', handler);
      this.cleanupFunctions.push(() => {
        window.removeEventListener('unhandledrejection', handler);
      });

      return true;
    } catch (error) {
      return false;
    }
  }

  /**
   * Method 2: Hermes Engine - Use HermesInternal API
   */
  private setupHermesTracking(): boolean {
    try {
      // Check if Hermes engine is available
      const hermesInternal = (global as any).HermesInternal;
      if (
        hermesInternal &&
        typeof hermesInternal.enablePromiseRejectionTracker === 'function'
      ) {
        hermesInternal.enablePromiseRejectionTracker({
          allRejections: true, // Track both handled and unhandled rejections
          onUnhandled: (id: number, error: any) => {
            this.onRejectionHandler(error);
          },
          onHandled: (id: number) => {
            // Promise was handled later - could be used for analytics in the future
          },
        });

        // Cleanup function (if available)
        this.cleanupFunctions.push(() => {
          // Hermes doesn't provide a disable method, but we can set a no-op handler
          if (hermesInternal?.enablePromiseRejectionTracker) {
            hermesInternal.enablePromiseRejectionTracker({
              allRejections: false,
              onUnhandled: () => {},
              onHandled: () => {},
            });
          }
        });

        return true;
      }
      return false;
    } catch (error) {
      return false;
    }
  }

  /**
   * Method 3: React Native Internal - Use promise/setimmediate/rejection-tracking for JSC
   */
  private setupReactNativeTracking(): boolean {
    try {
      // Try to access React Native's internal promise tracking module
      const tracking = require('promise/setimmediate/rejection-tracking');

      if (tracking && typeof tracking.enable === 'function') {
        tracking.enable({
          allRejections: true,
          onUnhandled: (id: number, error: any) => {
            this.onRejectionHandler(error);
          },
          onHandled: (id: number) => {
            // Promise was handled later - tracking for potential analytics
          },
        });

        this.cleanupFunctions.push(() => {
          // Try to disable tracking if possible
          if (tracking.disable && typeof tracking.disable === 'function') {
            tracking.disable();
          }
        });

        return true;
      }
      return false;
    } catch (error) {
      return false;
    }
  }

  /**
   * Clean up all tracking mechanisms
   */
  public cleanup(): void {
    // Clean up all tracking mechanisms and restore original state
    this.cleanupFunctions.forEach((cleanup) => {
      try {
        cleanup();
      } catch (error) {
        // Silently handle cleanup errors to avoid disrupting app shutdown
      }
    });
    this.cleanupFunctions = [];
  }
}

/**
 * Helper function to detect JavaScript engine (simplified for Web/Hermes/JSC)
 */
export function detectJavaScriptEngine(): 'web' | 'hermes' | 'jsc' {
  if (Platform.OS === 'web') {
    return 'web';
  }

  if (typeof (global as any).HermesInternal !== 'undefined') {
    return 'hermes';
  }

  // For non-Hermes engines on mobile, assume JSC (most common)
  return 'jsc';
}

/**
 * Public API for easy integration
 */
/**
 * Sets up advanced promise rejection tracking with automatic engine detection
 * @param onRejection Callback function to handle unhandled promise rejections
 * @returns AdvancedPromiseRejectionTracker instance for cleanup and testing
 */
export function setupAdvancedPromiseRejectionTracking(
  onRejection: PromiseRejectionHandler
): AdvancedPromiseRejectionTracker {
  const tracker = new AdvancedPromiseRejectionTracker(onRejection);
  tracker.setup();
  return tracker;
}
