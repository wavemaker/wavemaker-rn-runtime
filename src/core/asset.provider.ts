import React from 'react';

const AssetContext = React.createContext<(path: string) => number | string>(null as any);

export const AssetProvider = AssetContext.Provider;
export const AssetConsumer = AssetContext.Consumer;