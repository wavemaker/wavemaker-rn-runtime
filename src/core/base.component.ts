import { merge } from 'lodash';
import React, { ReactNode } from 'react';
import BASE_THEME, { DEFAULT_CLASS, DEFAULT_STYLE } from '../styles/theme';
import { PropsProvider } from './props.provider';

export interface ComponentState<T extends BaseProps> {
    props: T;
}

export class BaseProps {
    name: string = null as any;
    themeToUse = BASE_THEME;
    show = true;
    styles: any = null;
    onInit: Function = null as any;
    onDestroy: Function = null as  any;
}

export abstract class BaseComponent<T extends BaseProps> extends React.Component<T, ComponentState<T>> {
    public styles: any = DEFAULT_STYLE;
    private propertyProvider: PropsProvider<T>;
    public proxy: BaseComponent<T>;

    constructor(markupProps: T, public defaultClass = DEFAULT_CLASS, private defaultStyles = DEFAULT_STYLE, defaultProps?: T) {
        super(markupProps);
        this.propertyProvider = new PropsProvider<T>(
            Object.assign(defaultProps || {}, markupProps), 
            (name: string, $new: any, $old: any) => {
                this.onPropertyChange(name, $new, $old);
            });
        this.state = {
            props: this.propertyProvider.get()
        };
        //@ts-ignore
        this.proxy = (new Proxy(this, {
            get: (target, prop, receiver): any => {
                const propName = prop.toString();
                if (this.propertyProvider.has(propName)) {
                    //@ts-ignore
                    return this.state.props[propName];
                }
                return Reflect.get(target, prop, receiver);
            },
            set: (target: any, prop, value: any): any => {
                const propName = prop.toString();
                if (this.propertyProvider.has(propName)) {
                    // @ts-ignore
                    this.state.props[propName] = value;
                    this.setState({
                        props: this.propertyProvider.get()
                    })
                    return true;
                } else {
                    return Reflect.set(target, prop, value);
                }
            }
        }));
    }

    onPropertyChange(name: string, $new: any, $old: any) {

    }

    shouldComponentUpdate(nextProps: T, nextState: ComponentState<T>, nextContext: any) {
        if (this.propertyProvider.check(nextProps)) {
            this.setState({
                props: this.propertyProvider.get()
            });
            return true;
        }
        return !!(super.shouldComponentUpdate && super.shouldComponentUpdate(nextProps, nextState, nextContext));
    }

    componentDidMount() {
        this.props.onInit && this.props.onInit(this.proxy);
    }

    componentWillUnmount() {
        this.props.onDestroy && this.props.onDestroy(this);
    }

    invokeEventCallback(eventName: string, args: any[]) {
        //@ts-ignore
        const callBack: Function = this.props[eventName];
        if (callBack) {
            callBack.apply(this.proxy, args);
        }
        if (eventName === 'onTap') {
            this.invokeEventCallback('onClick', args);
        }
    }

    public render(): ReactNode {
        this.styles = merge({}, this.props.themeToUse.getStyle(this.defaultClass) || this.defaultStyles, this.props.styles);
        return null;
    }
}