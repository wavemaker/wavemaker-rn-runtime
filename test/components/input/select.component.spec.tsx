import React, { ReactNode, createRef } from 'react';
import { Text, TouchableOpacity } from 'react-native';
import WmSelect from '@wavemaker/app-rn-runtime/components/input/select/select.component';
import WmSelectProps from '../../../src/components/input/select/select.props';
import {
  render,
  screen,
  fireEvent,
  waitFor,
  cleanup,
} from '@testing-library/react-native';
import WmIcon from '@wavemaker/app-rn-runtime/components/basic/icon/icon.component';

import * as AccessibilityConfig from '../../../src/core/accessibility';

import AppModalService from '@wavemaker/app-rn-runtime/runtime/services/app-modal.service';
import { AssetProvider } from '@wavemaker/app-rn-runtime/core/asset.provider';
import { ModalProvider } from '@wavemaker/app-rn-runtime/core/modal.service';

const dataItems = [
  {
    name: 'name0',
    dataValue: 'dataValue0',
  },
  {
    name: 'name1',
    dataValue: 'dataValue1',
  },
  {
    name: 'name2',
    dataValue: 'dataValue2',
  },
];

const timer = (time = 100) =>
  new Promise((resolve: any, reject) => {
    setTimeout(() => resolve(), time);
  });

const loadAsset = (path) => path;

const defaultProps = {
  placeholder: 'Select an option',
  dataset: [
    {
      name: 'name0',
      dataValue: 'dataValue0',
    },
    {
      name: 'name1',
      dataValue: 'dataValue1',
    },
    {
      name: 'name2',
      dataValue: 'dataValue2',
    },
  ],
  datafield: 'name',
  displayfield: 'name',
};

AppModalService.modalsOpened = [];

function renderComponentWithWrappers(props = {}) {
  return render(
    <ModalProvider value={AppModalService}>
      <AssetProvider value={loadAsset}>
        <WmSelect
          {...defaultProps}
          dataset={dataItems}
          datafield="name"
          displaylabel="name"
          {...props}
        />
      </AssetProvider>
    </ModalProvider>
  );
}

describe('WmSelect', () => {
  beforeEach(() => {
    jest
      .spyOn(AccessibilityConfig, 'isScreenReaderEnabled')
      .mockReturnValue(false);
  });

  afterEach(() => {
    jest.clearAllMocks();
    cleanup();
  });

  // Check Rendering with Default Props
  it('renders correctly with default props', () => {
    render(<WmSelect {...defaultProps} name="select1" />);
    expect(screen.getByText('Select an option')).toBeTruthy();
  });

  // Disabled State
  it('does not allow selection when disabled', async () => {
    const tree = renderComponentWithWrappers({
      disabled: true,
      accessibilityrole: 'select',
    });
    const select = tree.getByText('Select an option');
    const selectParent = tree.getByRole('select');
    expect(selectParent.props.accessibilityState.disabled).toBe(true);
    fireEvent.press(select);

    await timer(200);

    const renderOptions = AppModalService.modalOptions;
    const Content = () => {
      return <>{renderOptions.content}</>;
    };
    const subTree = render(<Content />);
    expect(subTree.queryByText('name0')).toBe(null);
  });

  // Event Handling
  it('handles onFocus and onBlur events', async () => {
    const onFocusMock = jest.fn();
    const onBlurMock = jest.fn();
    const ref = createRef();
    const tree = renderComponentWithWrappers({
      onFocus: onFocusMock,
      onBlur: onBlurMock,
      ref,
    });
    const select = tree.getByText('Select an option');

    fireEvent.press(select);

    expect(onFocusMock).toHaveBeenCalled();

    await timer(200);

    const renderOptions = AppModalService.modalOptions;
    const Content = () => {
      return <>{renderOptions.content}</>;
    };
    const subTree = render(<Content />);

    const cancel = subTree.getByText('name0');
    fireEvent.press(cancel);
    await timer(500);

    expect(onBlurMock).toHaveBeenCalled();
  });

  // Dataset Handling
  it('handles dataset properly and renders options based on it', async () => {
    const tree = renderComponentWithWrappers({ name: 'select1' });
    const select = tree.getByText('Select an option');
    fireEvent.press(select);

    await timer(200);

    const renderOptions = AppModalService.modalOptions;
    const Content = () => {
      return <>{renderOptions.content}</>;
    };
    const subTree = render(<Content />);
    expect(subTree.getByText('name0')).toBeTruthy();
    expect(subTree.getByText('name1')).toBeTruthy();
    expect(subTree.getByText('name2')).toBeTruthy();
  });

  // Accessibility Props
  it('applies accessibility props correctly', () => {
    render(
      <WmSelect
        {...defaultProps}
        accessibilitylabel="Select an option"
        hint="wm-select"
        accessibilityrole="select"
      />
    );
    expect(screen.getByLabelText('Select an option')).toBeTruthy();
    expect(screen.getByRole('select')).toBeTruthy();
    expect(screen.getByA11yHint('wm-select')).toBeTruthy();
  });

  //default value
  it('renders default value correctly', async () => {
    const tree = renderComponentWithWrappers({
      accessibilityrole: 'select',
      datavalue: 'name1',
    });

    const select = tree.getByText('name1');
    fireEvent.press(select);

    await timer(200);

    const renderOptions = AppModalService.modalOptions;
    const Content = () => {
      return <>{renderOptions.content}</>;
    };
    const subTree = render(<Content />);
    expect(
      subTree.UNSAFE_getAllByType(WmIcon)[2].props.styles.root.opacity
    ).toBe(1);
  });

  // Item Selection
  it('selects item correctly and updates value', async () => {
    const ref = createRef();
    const onChangeEventMock = jest.fn();
    const tree = renderComponentWithWrappers({
      accessibilityrole: 'select',
      ref,
      onChange: onChangeEventMock,
    });
    const onChangefnMock = jest.spyOn(WmSelect.prototype, 'onChange');
    const select = tree.getByText('Select an option');
    const selectParent = tree.getByRole('select');

    expect(tree).toMatchSnapshot();
    fireEvent.press(select);

    await timer(200);

    const renderOptions = AppModalService.modalOptions; //[0];
    const Content = () => {
      return <>{renderOptions.content}</>;
    };
    const subTree = render(<Content />);
    expect(subTree).toMatchSnapshot();
    const option = subTree.getByText('name0');

    fireEvent.press(option);

    await timer(1000);

    expect(onChangeEventMock).toHaveBeenCalled();
    expect(onChangefnMock).toHaveBeenCalled();
    console.log(ref.current.state);
    expect(tree).toMatchSnapshot();
    await waitFor(() => {
      expect(tree.getByText('name0')).toBeTruthy();
    });
  });

  // Error Handling
  it('handles validation and error states correctly', async () => {
    const triggerValidationMock = jest.fn();
    const ref = createRef();
    const validateMock = jest.spyOn(WmSelect.prototype, 'validate');
    // const validateMock = jest.spyOn(, 'validateField');
    const tree = renderComponentWithWrappers({
      triggerValidation: triggerValidationMock,
      datavalue: '',
      ref,
      required: true,
    });
    const select = screen.getByText('Select an option');
    fireEvent.press(select);

    await timer(200);

    const renderOptions = AppModalService.modalOptions; //[0];
    const Content = () => {
      return <>{renderOptions.content}</>;
    };
    const subTree = render(<Content />);
    renderOptions.onClose();
    await timer(200);

    expect(triggerValidationMock).toHaveBeenCalled();
    expect(validateMock).toHaveBeenCalled();
  });

  // show property
  it('handles show property correctly', () => {
    render(
      <WmSelect
        {...defaultProps}
        accessibilitylabel="Select an option"
        show={false}
      />
    );
    const searchParent = screen.getByLabelText('Select an option');
    expect(searchParent.props.style[0].width).toBe(0);
    expect(searchParent.props.style[0].height).toBe(0);
  });
});
