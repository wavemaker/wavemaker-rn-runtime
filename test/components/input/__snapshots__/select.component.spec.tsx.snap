// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WmSelect renders correctly with default props 1`] = `
[
  <View
    onLayout={[Function]}
    style={
      {
        "height": 0,
        "width": 0,
      }
    }
  />,
  <View>
    <View
      onLayout={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderColor": "#dddddd",
            "borderRadius": 6,
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "row",
            "minWidth": 160,
            "paddingBottom": 12,
            "paddingLeft": 12,
            "paddingRight": 12,
            "paddingTop": 12,
          },
          {},
          {
            "backgroundColor": "#ffffff",
          },
          {},
        ]
      }
    >
      <Text
        accessibilityState={
          {
            "disabled": false,
          }
        }
        accessible={true}
        onPress={[Function]}
        style={
          [
            {
              "color": "#151420",
              "flex": 1,
              "fontFamily": "Roboto",
              "fontSize": 16,
              "userSelect": "text",
            },
            {
              "color": "#bbbbbb",
            },
          ]
        }
        testID="select1_i"
      >
        Select an option
      </Text>
      <View
        onLayout={[Function]}
        style={
          {
            "height": 0,
            "width": 0,
          }
        }
      />
      <View
        style={
          [
            {
              "alignItems": "center",
              "alignSelf": "center",
              "backgroundColor": "transparent",
              "borderRadius": 0,
              "justifyContent": "center",
              "minHeight": 0,
              "paddingBottom": 6,
              "paddingLeft": 6,
              "paddingRight": 0,
              "paddingTop": 6,
              "rippleColor": "transparent",
            },
            {
              "flexDirection": "column",
              "overflow": "hidden",
              "paddingBottom": 0,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 0,
            },
            {},
          ]
        }
        testID="non_animatableView"
      >
        <View
          accessibilityLabel={null}
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "height": null,
                "justifyContent": "center",
                "paddingBottom": 6,
                "paddingLeft": 6,
                "paddingRight": 0,
                "paddingTop": 6,
                "width": "100%",
              },
            ]
          }
          testID="undefined_a"
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                {
                  "flexDirection": "row",
                },
              ]
            }
          >
            <View
              onLayout={[Function]}
              style={
                {
                  "height": 0,
                  "width": 0,
                }
              }
            />
            <View>
              <View
                style={
                  {
                    "alignItems": "center",
                    "alignSelf": "auto",
                    "flexDirection": "row",
                    "paddingLeft": -8,
                    "paddingRight": -8,
                  }
                }
                testID="non_animatableView"
              >
                <View>
                  <View>
                    <Text
                      accessible={false}
                      size={16}
                      style={
                        [
                          {
                            "color": undefined,
                          },
                          {
                            "paddingLeft": 0,
                            "paddingRight": 8,
                          },
                          {
                            "transform": [
                              {
                                "rotate": "0deg",
                              },
                            ],
                          },
                        ]
                      }
                      testID="undefined_icon_icon"
                    >
                      keyboard-arrow-down
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>,
]
`;

exports[`WmSelect should render skeleton with respect to root styles when show skeleton is true 1`] = `
[
  <View
    onLayout={[Function]}
    style={
      {
        "height": 0,
        "width": 0,
      }
    }
  />,
  <View>
    <View
      onLayout={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderColor": "#dddddd",
            "borderRadius": 6,
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "row",
            "minWidth": 160,
            "paddingBottom": 12,
            "paddingLeft": 12,
            "paddingRight": 12,
            "paddingTop": 12,
          },
          {},
          {
            "backgroundColor": "#ffffff",
          },
          {
            "justifyContent": "space-between",
          },
        ]
      }
    >
      <View
        onLayout={[Function]}
        style={
          {
            "height": 0,
            "width": 0,
          }
        }
      />
      <View
        onLayout={[Function]}
        style={
          {
            "backgroundColor": "#eeeeee",
            "borderRadius": 8,
            "height": 16,
            "overflow": "hidden",
            "width": 100,
          }
        }
      />
      <View
        onLayout={[Function]}
        style={
          {
            "height": 0,
            "width": 0,
          }
        }
      />
      <View
        onLayout={[Function]}
        style={
          {
            "backgroundColor": "#eeeeee",
            "borderRadius": 32,
            "height": 48,
            "overflow": "hidden",
            "width": 96,
          }
        }
      />
    </View>
  </View>,
]
`;
